### ISSUE TITLE
Tree View problems with Category Id collision

### ISSUE THREAD

**Issue by Tschrock** [SparkDevNetwork/Rock#1632]
We've been having problems selecting workflows in a workflow tree view (For ex: actions to show on a person's profile) and I've found that it is due to having a workflow and a category with the same Id. In our case this is preventing us from expanding or selecting anything in the conflicting category. This may also affect the schedule tree view as I know we've had some issues with them before?

**Comment by arranf** [comment]
@Tschrock Hey can you go into more detail on this please? Some steps to reproduce would be great!

**Comment by Tschrock** [comment]
Just checked in v5.2 and this seems to be partially fixed.

To try to illustrate:

![capture28_](https://cloud.githubusercontent.com/assets/2406499/17626318/c010a4ae-607a-11e6-8cd7-5818256765a9.PNG)

The original problem was that if you opened the tree view with the workflow selected, you could not expand the category - that seems to be fixed. However, now if you try to unselect one of them (either the workflow or the category), they will both still be selected when you re-open the settings.

**Comment by mikejed** [comment]
This is very difficult to reproduce, since the collision of those numbers relies on a very large number of categories/dataviews to be created (> 1000 at least), given where the demo server presently is in it's index number of each.

However, looking at the HTML, I see that categories use
```
<li class="rocktree-item rocktree-leaf" data-id="C6176" data-iscategory="true" data-entityid="6176">
```

and workflows use
```
<li class="rocktree-item rocktree-leaf" data-id="2019" data-iscategory="false" data-entityid="2019">
```

So, it's easy to see how the JS implementation of the treeview could get confused with the similar formats of the `data-entityid` values.

Perhaps changing the category data-entityid to begin with `C` like the data-id does, would resolve the issue (if indeed it's still an issue). Or perhaps the treeview script (which I looked for but can't find) can be updated to double-check `data-iscategory` and perform the correct actions based on what the item clicked is.

@Tschrock do you still happen to have colliding keys in your installation? 😇

**Comment by Tschrock** [comment]
I don't know if we still have that conflicting key, but I've gone onto to prealpha server and purposefully created a conflicting category Id using the following SQL:
```sql
SET IDENTITY_INSERT Category ON;
INSERT INTO Category (Id, IsSystem, EntityTypeId, Name, Guid, [Order])
VALUES (1017, 0, 32, 'Id Test Category', NEWID(), 0);
SET IDENTITY_INSERT Category OFF;
```
(The Id 1017 here matches the Id of the DISC Request workflow on that server)

With that in place, I went on the person's profile page and attempted to edit the Workflow Actions in the Bio block. 

There were two issues that I observed:
1. It is impossible to expand the Category that has the conflicting Id.
2. Selecting either the Category or the Workflow Type and saving will have them both selected upon reopening.

**Comment by arranf** [comment]
Here's some SQL to reproduce locally on a fresh DB

```
SET IDENTITY_INSERT Category ON;
INSERT INTO Category (Id, IsSystem, ParentCategoryId, EntityTypeId, EntityTypeQualifierColumn, [Name], IconCssClass, [Guid], [Order])
VALUES (195, 0, NULL, (SELECT TOP 1 [Id] FROM [EntityType] WHERE [Name] = 'Rock.Model.WorkflowType'), NULL, 'Test Workflow Category', '', NEWID(), 0);
SET IDENTITY_INSERT Category OFF;
```

```
SET IDENTITY_INSERT WorkflowType ON;
INSERT INTO WorkflowType (Id, IsSystem, IsActive, Name, Description, CategoryId, [Order], WorkTerm, ProcessingIntervalSeconds, IsPersisted, LoggingLevel, [Guid])
VALUES (195, 0, 1, 'Test Workflow', '', 59, 0, '', NULL, 0, 0, NEWID());
SET IDENTITY_INSERT WorkflowType OFF;
```

**Comment by arranf** [comment]
Looks like `rockTree.js` and `itemPicker.js` were written under the assumption that either a control would not represent multiple entities (i.e. not categories and workflows).

Ids are stored in a `selectedIds` array and then nodes are marked selected by fetching them using this id. Possible solutions would be creating a separate mechanism for making categories as selected (with their own array) or fully qualifying ids with their entity name (which would create a unique id).

**Comment by azturner** [comment]
We are running into this issue more and more often. For example in this example, we cannot select the needed 'Creative Arts Project Request' workflow because it happens to have same id as the 'Check-in' category (59). 

![image](https://user-images.githubusercontent.com/548180/73641380-2a937180-462d-11ea-8cf1-435aadf28cb7.png)

**Comment by cabal95** [comment]
Much of this seems to stem from the ability of the ItemPicker control's `AllowCategorySelection` property. Which as far as I can tell from a code search is _only_ used by the CategoryPicker subclass - in which case the only things listed are the categories and not the items.

By reading the documentation on this property the original intent was to allow the selection of _both_ categories and items, but that will never work out because of Id conflicts. The control only has a single `ItemIds` property to retrieve the selected identifiers, with no way to know if that is a category id or an item id.

## Proposed Solution

TL;DR; change the functionality of ItemPicker so that you can select Items **OR** you can select categories, but not both.

1. Add a new property to ItemPicker called `UseCategorySelection`. This would change the behavior of the ItemPicker to allow _only_ categories to be selected.

2. Deprecate `ItemPicker.AllowCategorySelection` for clarify sake, this would become an alias to the `UseCategorySelection` property above for now.  It could then be completely removed in Rock v13.

3. Modify the `CategoryPicker` subclass to use the new property name.

4. Modify the `rockTree.js` and `itemPicker.js` files to prefix category identifiers in the DOM with a `C`. This `C` would be stripped before being sent back to the `ItemPicker` control and automatically prepended when doing initialization (if in category mode).

### Testing

I have mocked these changes up on my dev instance and they seem to solve the problem without breaking any existing (core) functionality.

### Plugins

Any plugins that use the existing `AllowCategorySelection` with the express intent to allow selection of both categories and items would essentially break - but they are a ticking time bomb anyway.

**Comment by cabal95** [comment]
## Proposed Solution 2.0

TL;DR; change the functionality of ItemPicker so that you can select Items OR you can select categories, but not both. This would only affect the `ItemPicker` control, not any blocks that utilize `rockTree.js` directly.

1. Add new initialization parameter to `rockTree.js` called `categoryPrefix` which defaults to an empty string.

2. In the default item mapping function of `rockTree.js`, check if the item `isCategory` is true and if so, prepend the `categoryPrefix` string. This means a workflow id of 123 would end up as `123` in the DOM and a category id of 123 would end up as `C123` (for example if the `categoryPrefix` was set to `C`) in the DOM.

3. Add a new property to ItemPicker called `UseCategorySelection`. This would change the behavior of the ItemPicker to allow only categories to be selected.

4. Deprecate ItemPicker `AllowCategorySelection` for clarity sake, this would become an alias to the `UseCategorySelection` property above for now. It could then be completely removed in Rock v13.

5. When ItemPicker `UseCategorySelection` is true:
    a. The new `categoryPrefix` setting will be set to `C`
    b. Any `ItemIds` set in the ItemPicker will have the `C` prefix added before being passed to JavaScript.
    c. When reading the `ItemIds` in the ItemPicker, any `C` prefixes will be trimmed.

6. Modify the CategoryPicker subclass to use the new property name.

### How this affects existing functionality

Items 1 and 2 together ensure that anything using `rockTree.js` directly do not have any change in functionality by default (i.e. 3rd party plugins).

Item 6, `CategoryPicker`, seems to be the only place the `AllowCategorySelection` property is used.

Existing blocks that use `rockTree.js` directly will need to be double-tested to verify they still work as intended. These blocks are:

* Cms > Page Map (should be fine since it doesn't use categories)
* Core > Category Tree View (uses custom mapping already which is similar to the above)
* Core > Location Tree View (should be fine since I don't think it uses categories)
* Groups > Group Tree View (should be fine since it doesn't use categories)
* Utility > Html Editor File Browser (should be fine since it doesn't use categories)
* Asset Manager Blocks (assetManager.js) (I would assume also should be fine since it probably doesn't use categories)

### Pre-Testing

I have mocked the important parts of these changes up on my dev instance and they seem to solve the problem. The Workflow Configuration screen's tree view (Core > Category Tree View) also works with this change. The other blocks mentioned above would need testing to ensure they also still work, but I don't think any of them use categories.

Obviously this is a potentially big change so once implemented it would need lots of user testing to ensure nothing broke as a side affect.

### Plugins

Any plugins that use the existing AllowCategorySelection with the express intent to allow selection of both categories and items would essentially break - but they are a ticking time bomb anyway.

**Comment by nairdo** [comment]
Daniel @cabal95, if you're willing to PR your mock to the develop branch, we can get it tested and then picked into the 10.3 branch if all is well.  Otherwise we'll just use your described idea and code up the solution for a future release.

**Comment by cabal95** [comment]
Will do, I'll work on that this week.

**Comment by cabal95** [comment]
@nairdo Created PR against develop branch, #4098.
